certora/specs/ConsistentState.spec

```diff
@@ -155,5 +170,92 @@ function hasSupplyCapIsEnabled(MetaMorphoHarness.Id id) returns bool {
     return supplyCap > 0 => enabled;
 }
 
+// Check that having a positive supply cap implies that the market is enabled.
+// This invariant is useful to conclude that market that are not enabled cannot be interacted with (notably for reallocate).
 invariant supplyCapIsEnabled(MetaMorphoHarness.Id id)
     hasSupplyCapIsEnabled(id);
+
+function hasDistinctIdentifiers(uint256 i, uint256 j) returns bool {
+    return i != j => withdrawQueue(i) != withdrawQueue(j);
+}
+
+// Check that there are no duplicate markets in the withdraw queue.
+invariant distinctIdentifiers(uint256 i, uint256 j)
+    hasDistinctIdentifiers(i, j)
+{
+    preserved updateWithdrawQueue(uint256[] indexes) with (env e) {
+        require hasDistinctIdentifiers(indexes[i], indexes[j]);
+    }
+}
+
+function isInWithdrawQueueIsEnabled(uint256 i) returns bool {
+    if(i >= withdrawQueueLength()) return true;
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
+    bool enabled;
+    _, enabled, _ = config(id);
+
+    return enabled;
+}
+
+// Check that markets in the withdraw queue are enabled.
+invariant inWithdrawQueueIsEnabled(uint256 i)
+    isInWithdrawQueueIsEnabled(i)
+filtered {
+    f -> f.selector != sig:updateWithdrawQueue(uint256[]).selector
+}
+
+rule inWithdrawQueueIsEnabledPreservedUpdateWithdrawQueue(env e, uint256 i, uint256[] indexes) {
+    uint256 j;
+    require isInWithdrawQueueIsEnabled(indexes[i]);
+
+    requireInvariant distinctIdentifiers(indexes[i], j);
+
+    updateWithdrawQueue(e, indexes);
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
+    // Safe require because j is not otherwise constrained.
+    // The ghost variable deletedBy is useful to make sure that markets are not permuted and deleted at the same time in updateWithdrawQueue.
+    require j == deletedBy(id);
```

Comment:
I don't get why we need this and why we even need j

Reply:
`j` is here to make sure that there are no duplicates of `id` in the withdraw queue that are removed (because that would mean setting the `enabled` flag to `false` which breaks the property). We know that there are no duplicates at all in the withdraw queue, thanks to the invariant `distinctIdentifiers`, but we can't use `forall` here so we must specify which indexes lead to different identifiers.

The trick is the following:
- we want to specify that there are no identifier that is a duplicate of `id` and that is removed
- if there is one duplicate (or more) that is removed, then then `j` would be an index that correspond to such an identifier in the initial withdraw queue
- we specify that `j` and `indexes[i]` correspond to different identifiers in the initial withdraw queue, which is what we want to specify because `indexes[i]` corresponds to `id` in the initial withdraw queue

Reply:
get it